File: java_solutions_1/524.longest-word-in-dictionary-through-deleting.java
/*
 * @lc app=leetcode id=524 lang=java
 *
 * [524] Longest Word in Dictionary through Deleting
 */

class Solution {
    public String findLongestWord(String s, java.util.List<String> dictionary) {
        return dictionary.stream().sorted((a, b) -> a.length() != b.length() ? Integer.compare(b.length(), a.length()) : a.compareTo(b)).filter(w -> java.util.stream.Stream.of(new int[]{0}).map(ptr -> s.chars().anyMatch(c -> ptr[0] < w.length() && c == w.charAt(ptr[0]) && ++ptr[0] == w.length())).findFirst().get()).findFirst().orElse("");
    }
}

File: kotlin_solutions/524.longest-word-in-dictionary-through-deleting.kt
/*
 * @lc app=leetcode id=524 lang=kotlin
 *
 * [524] Longest Word in Dictionary through Deleting
 */

class Solution { fun findLongestWord(s: String, d: List<String>) = d.fold("") { longest, word -> word.fold(0 to 0) { (idxS, idxW), _ -> generateSequence(idxS) { it + 1 }.takeWhile { it < s.length }.firstOrNull { s[it] == word.getOrNull(idxW) }?.let { (it + 1) to (idxW + 1) } ?: (s.length to idxW) }.let { (_, idxW) -> when { idxW == word.length && (word.length > longest.length || (word.length == longest.length && word < longest)) -> {word} else -> longest } } } }